scripts/release-notes.py

```diff
@@ -395,68 +472,94 @@ def analyze_pr(merge, pr):
+            continue
+        seen_commits.add(commit)
 
         if not commit.message.startswith("Merge"):
-            extract_release_notes(pr, note, commit)
-
+            missing_item, prauthors = extract_release_notes(pr, note, commit)
+            authors.update(prauthors)
             ncommits += 1
-            author = author_aliases.get(commit.author.name, commit.author.name)
-            if author != 'GitHub':
-                authors.add(author)
-            committer = author_aliases.get(commit.committer.name, commit.committer.name)
-            if committer != 'GitHub':
-                authors.add(committer)
-
-        # Exclude any parents reachable from the other side of the
-        # PR merge commit.
-        for parent in commit.parents:
-          if not repo.is_ancestor(parent, merge.parents[0]):
-            commits_to_analyze.append(parent)
+            if missing_item is not None:
+                missing_items.append(missing_item)
 
+        for parent in commit.parents:
+            if not repo.is_ancestor(parent, merge.parents[0]):
+                # We're not yet back on the main branch. Just continue digging.
+                commits_to_analyze.append(parent)
+            else:
+                # The parent is on the main branch. We're done digging.
+                # print("found merge parent, stopping. final authors", authors)
+                pass
+
+    if ncommits == len(missing_items):
+        # None of the commits found had a release note. List them.
+        for item in missing_items:
+            missing_release_notes.append(item)
+            
     text = repo.git.diff(merge_base.hexsha, tip.hexsha, '--', numstat=True)
     stats = Stats._list_from_string(repo, text)
 
-    individual_authors.update(authors)
+    collect_item(pr, note, merge.hexsha[:shamin], ncommits, authors, stats.total)
 
+def collect_item(pr, prtitle, sha, ncommits, authors, stats):
+    individual_authors.update(authors)
     if len(authors) == 0:
         authors.add("Unknown Author")
-
     item = {
-        'title': note,
+        'title': prtitle,
         'pr': pr,
-        'sha': merge.hexsha[:shamin],
+        'sha': sha,
         'ncommits': ncommits,
         'authors': ", ".join(sorted(authors)),
-        'insertions': stats.total['insertions'],
-        'deletions': stats.total['deletions'],
-        'files': stats.total['files'],
-        'lines': stats.total['lines'],
+        'insertions': stats['insertions'],
+        'deletions': stats['deletions'],
+        'files': stats['files'],
+        'lines': stats['lines'],
         }
 
     history = per_group_history.get(item['authors'], [])
     history.append(item)
     per_group_history[item['authors']] = history
 
+def analyze_standalone_commit(commit):
+    # Some random out-of-branch commit. Let's not forget them.
+    authors = collect_authors(commit)
+    title = commit.message.split('\n')[0].strip()
+    item = {'authors': ', '.join(sorted(authors)),
```

Comment:
ok i've seen this little snippet enough times that i'll suggest it get extracted